File: 148_Sort List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def findMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = fast = head
        while fast and fast.next:
            pre = slow
            slow = slow.next
            fast = fast.next.next
        pre.next = None
        return slow
    
    def merge(self, left: Optional[ListNode], right: Optional[ListNode]) -> Optional[ListNode]:
        vHead = cur = ListNode()

        while left and right:
            if left.val < right.val:
                cur.next = left
                left = left.next
            else:
                cur.next = right
                right = right.next
            cur = cur.next
        
        cur.next = left if left else right
        return vHead.next

    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # 0. case: <= 1 node
        if not (head and head.next): return head
        # 1. divide: find middle & split into two halves
        right = self.findMiddle(head)
        left = head

        # 2. sort
        left = self.sortList(left)
        right = self.sortList(right)

        # 3. merge / conquer
        return self.merge(left, right)

File: 148_Sort List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode findMiddle(ListNode head) {
        ListNode slow = head, fast = head, prev = null;

        while (fast != null && fast.next != null) {
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }
        prev.next = null;
        return slow;
    }

    private ListNode merge(ListNode left, ListNode right) {
        ListNode vHead = new ListNode();
        ListNode cur = vHead;

        while (left != null && right != null) {
            if (left.val < right.val) {
                cur.next = left;
                left = left.next;
            } else {
                cur.next = right;
                right = right.next;
            }
            cur = cur.next;
        }

        cur.next = (left != null) ? left : right;
        return vHead.next;
    }

    public ListNode sortList(ListNode head) {
        // 1. case: <= 1 node
        if (head == null || head.next == null) return head;

        // 2. divide: split into two halves
        ListNode right = findMiddle(head);
        ListNode left = head;

        // 3. sort
        left = sortList(left);
        right = sortList(right);

        // 4. merge / conquer
        return merge(left, right);
    }
}